src/gleam/regex.gleam

```diff
@@ -3,8 +3,26 @@
 //// all of the PCRE library is interfaced and some parts of the library go beyond
 //// what PCRE offers. Currently PCRE version 8.40 (release date 2017-01-11) is used.
 
+import gleam/option.{Option}
+
 pub external type Regex
 
+/// The details about a particular match:
+///
+/// - match — the full string of the match.
+/// - index — the index of the match in the original string.
+/// - number — each match is numbered, starting from 1
+/// - submatches — a Regex can have subpatterns, sup-parts that are in parentheses.
+///
+pub type Match {
+  Match(
+    match: String,
+    index: Int,
+    number: Int,
+    submatches: List(Option(String)),
```

Comment:
Can this not be just a list of Strings? an empty list, when no submatches

Reply:
I followed Elm's way of doing this. I assume Elm did that for easier pattern matching? (`None` vs empty string test).

Reply:
Could you give some examples of what would result in what values? I don't understand what it is for at present.

Reply:
Not sure what example would be helpful, but what it comes down to is that given the same regex, subexpression matches can be either present or absent.
In other words it handles the optionality of subexpressions `(...)?`.

Example:

```rust
  assert Ok(re) = regex.from_string("^([^:]+?)(:\\d+)?/(.*)")


  regex.scan(re, "foo.com:80/bar/baz")
  |> should.equal(
    [
      Match(
        match: "foo.com:80/bar/baz",
        index: 0,
        submatches: [Some("foo.com"), Some(":80"), Some("bar/baz")],
      ),
    ],
  )

  regex.scan(re, "foo.com/bar/baz")
  |> should.equal(
    [
      Match(
        match: "foo.com/bar/baz",
        index: 0,
        submatches: [Some("foo.com"), None, Some("bar/baz")],
      ),
    ],
  )
```

Reply:
> In other words it handles the optionality of subexpressions (...)?.

I've come to realize there is more to it than that, currently a zero-length match would also become `None`:

```rust
  regex.scan(re, "foo.com/")
  |> should.equal(
    [
      Match(
        match: "foo.com/",
        index: 0,
        submatches: [Some("foo.com"), None, None],
      ),
    ],
  )
```

Reply:
Ah! That makes sense, thank you.

Kinda shows how regular expressions are not the best fit for typed langauges imo, something like a parser generator would provide much more precise types.